src/Microsoft.Extensions.OData.Migration/ODataMigrationMiddleware.cs

```diff
@@ -0,0 +1,170 @@
+            string translatedPathAndQuery = TranslateUri(fullRequestUri).PathAndQuery;
+            if (context.Request.Query.Any())
+            {
+                string[] pathAndQuery = translatedPathAndQuery.Split('?');
+                context.Request.Path = new PathString(pathAndQuery[0]);
+                context.Request.QueryString = new QueryString("?" + pathAndQuery[1]);
+            }
+            else
+            {
+                context.Request.Path = new PathString(translatedPathAndQuery);
+            }
+        }
+
+        /// <summary>
+        /// Accept a V3 request URI and return a V4 request URI.  V4 request URI retains base path.
+        /// </summary>
+        /// <param name="requestUri">V3 Request URI</param>
+        /// <returns>V4 Request URI</returns>
+        public Uri TranslateUri(Uri requestUri)
+        {
+            // Use UriTranslator to walk v3 segments, translating each to v4 and returning.
+            ODataPath v3path = new ODataUriParser(this.v3Model, this.serviceRoot).ParsePath(requestUri);
+            UriSegmentTranslator uriTranslator = new UriSegmentTranslator(this.v4Model);
+            Microsoft.OData.UriParser.ODataPath v4path = new Microsoft.OData.UriParser.ODataPath(v3path.WalkWith(uriTranslator));
+
+            // Parse query options for translation
+            NameValueCollection queryNvc = HttpUtility.ParseQueryString(requestUri.Query);
+            Dictionary<string, string> queryOptions = queryNvc.AllKeys.ToDictionary(k => k.Trim(), k => queryNvc[k].Trim());
+
+            // Create a v4 ODataUri and utilized ODataUriExtensions methods to build v4 URI
+            Microsoft.OData.ODataUri v4Uri = new Microsoft.OData.ODataUri()
+            {
+                Path = v4path,
+                Filter = ParseFilterFromQueryOrNull(queryOptions, v4path, v3path)
+            };
+            Uri v4RelativeUri = Microsoft.OData.ODataUriExtensions.BuildUri(v4Uri, Microsoft.OData.ODataUrlKeyDelimiter.Parentheses);
+            Uri v4TranslatedUri = new Uri(this.serviceRoot, v4RelativeUri);
+
+            // Translate Query
+            if (queryOptions.ContainsKey("$filter"))
+            {
+                queryOptions["$filter"] = HttpUtility.ParseQueryString(v4TranslatedUri.Query)["$filter"];
+            }
+            if (queryOptions.ContainsKey("$inlinecount"))
+            {
+                queryOptions["$count"] = ParseInlineCountFromQuery(queryOptions["$inlinecount"]);
+                queryOptions.Remove("$inlinecount");
+            }
+
+            // Join and append query string if applicable
+            string v4Query = (queryOptions.Count > 0 ? "?" : "") + String.Join("&", queryOptions.Select(x => x.Key + "=" + x.Value).ToArray());
+            v4TranslatedUri = new Uri(Uri.UnescapeDataString(v4TranslatedUri.GetLeftPart(UriPartial.Path) + v4Query));
+
+            return v4TranslatedUri;
+        }
+
+        /// <summary>
+        /// 

```

Comment:
No summary. For private functions, I think it's fine not to follow the "verbose" logging format if not necessary. Use // followed by a brief description.